File: src/asset_director/asset_preview.py
"""Bounded, private inspection copies. No writes to the original library/project."""
from pathlib import Path
import copy
import shutil
from .core import Asset, Library, atomic_json, fields, file_hash, load_json, require, within

FORMATS = {'.blend', '.gltf', '.glb', '.fbx', '.bvh'}
MAX_BYTES = 512 * 1024 * 1024
MAX_FILES = 4096


def snapshot(request_file):
    request_file = Path(request_file).resolve()
    request = load_json(request_file, 2 * 1024 * 1024)
    fields(request, {'schema', 'id', 'title', 'version', 'source_kind', 'root', 'files', 'file', 'motion'},
           {'schema', 'id', 'title', 'version', 'source_kind', 'root', 'files', 'file'})
    require(request['schema'] == 'asset-director.asset-preview/1', 'INVALID_PREVIEW', 'Unknown preview request')
    records = request['files']
    require(isinstance(records, list) and 0 < len(records) <= MAX_FILES, 'RESOURCE_LIMIT', 'Preview supports at most 4096 package files')
    names = set()
    for f in records:
        fields(f, {'path', 'size', 'sha256'}, {'path', 'size', 'sha256'})
        require(type(f['size']) is int and f['size'] >= 0, 'INVALID_PREVIEW', 'Invalid file size')
        require(f['path'].casefold() not in names, 'INVALID_PREVIEW', 'Duplicate package member')
        names.add(f['path'].casefold())
    total = sum(f['size'] for f in records)
    require(total <= MAX_BYTES, 'RESOURCE_LIMIT', 'Preview copy exceeds 512 MiB; choose a smaller reviewed package')
    require(request['file'] in [f['path'] for f in records] and Path(request['file']).suffix.lower() in FORMATS,
            'FORMAT_UNSUPPORTED', 'Choose one recorded blend, glTF, GLB, FBX or BVH member')
    source_root = Path(request['root']).resolve(strict=True)
    directory = request_file.parent
    require(not directory.is_relative_to(source_root) and not source_root.is_relative_to(directory),
            'ORIGINAL_OVERWRITE', 'Preview output must be separate from its source root')
    destination = directory / 'library'
    require(not destination.exists(), 'PREVIEW_EXISTS', 'Preview attempts are immutable; request a new copy')
    require(shutil.disk_usage(directory).free >= total * 2 + 128 * 1024 * 1024,
            'DISK_SPACE', 'Not enough free space for this bounded preview')
    # Validate every byte before allocating the snapshot; verify again after copy.
    sources = []
    for f in records:
        p = within(source_root, f['path'])
        require(p.is_file() and p.stat().st_size == f['size'] and file_hash(p) == f['sha256'],
                'STALE_SOURCE', 'Source changed; refresh its catalog/package record')
        sources.append(p)
    destination.mkdir()
    files = []
    for source, f in zip(sources, records):
        relative = 'incoming/package/' + f['path']
        target = within(destination, relative)
        target.parent.mkdir(parents=True, exist_ok=True)
        with source.open('rb') as inp, target.open('xb') as out:
            shutil.copyfileobj(inp, out, 1024 * 1024)
        require(target.stat().st_size == f['size'] and file_hash(target) == f['sha256']
                and source.stat().st_size == f['size'] and file_hash(source) == f['sha256'],
                'STALE_SOURCE', 'Source changed during preview snapshot; failed attempt retained')
        files.append({**f, 'path': relative})
    # Inspection-only transient record in a separate SQLite library. It is not
    # intake into the user's catalog and carries no manufactured rights grant.
    meta = {'preview_only': True}
    motion = copy.deepcopy(request.get('motion') or {})
    if motion:
        fields(motion, {'file', 'action', 'slot', 'source_object', 'fps', 'frame_start', 'frame_end'},
               {'file', 'action', 'source_object', 'fps', 'frame_start', 'frame_end'})
        require(motion['file']['path'] == request['file'], 'INDEX_SOURCE_MISMATCH', 'Preview must use the indexed take member')
        motion['file'] = next(f for f in files if f['path'] == 'incoming/package/' + request['file'])
        meta.update(motion)
    asset = Asset('local', request['id'] + ':' + request['version'], request['title'],
                  'animation' if motion else 'model', '', local_files=files, metadata=meta)
    with Library(destination) as lib:
        lib.put(asset)
    return request, destination, asset


def prepare(request_file, blender):
    from . import jobs
    request_file = Path(request_file).resolve()
    receipt = request_file.parent / 'receipt.json'
    status = {'schema': 'asset-director.asset-preview/1', 'state': 'PREPARING',
              'selection_changed': False, 'production_use_approved': False}
    atomic_json(receipt, status)
    try:
        request, directory, asset = snapshot(request_file)
        status.update(source_id=request['id'], source_version=request['version'],
                      source_kind=request['source_kind'], title=request['title'], file=request['file'])
        with Library(directory) as lib:
            job = jobs.prepare(lib, 'asset-preview', asset_id=asset.id,
                               options={'file': 'incoming/package/' + request['file']})
            status.update(job_id=job['id'], state='RUNNING')
            atomic_json(receipt, status)
            job = jobs.run(lib, job['id'], blender, timeout=180)
            require(job['state'] == 'SUCCEEDED', 'PREVIEW_FAILED', 'Blender preview failed; inspect the retained job log')
            blend = next(f for f in job['outputs'] if f['path'].endswith('/result.blend'))
            report = next(f for f in job['outputs'] if f['path'].endswith('/result.json'))
            data = load_json(lib.verify_file(report))['data']
            viewer=request_file.parent/'PREVIEW_COPY.blend'
            with lib.verify_file(blend).open('rb') as inp,viewer.open('xb') as target:
                shutil.copyfileobj(inp,target,1024*1024)
            require(file_hash(viewer)==blend['sha256'],'STALE_PREVIEW','Viewer copy differs from the prepared job')
            # Recheck originals after the complete native operation.
            for f in request['files']:
                p = within(Path(request['root']), f['path'])
                require(p.stat().st_size == f['size'] and file_hash(p) == f['sha256'],
                        'STALE_SOURCE', 'Original changed during preparation; preview not launched')
            status.update(state='READY', blend={**blend, 'path': viewer.name},
                          job_blend={**blend,'path':'library/'+blend['path']},
                          report={**report, 'path': 'library/' + report['path']}, data=data)
            atomic_json(receipt, status)
            return status
    except Exception as error:
        status.update(state='FAILED', code=getattr(error, 'code', type(error).__name__),
                      message=str(error) if hasattr(error, 'code') else 'Preview preparation failed; retained attempt needs inspection')
        atomic_json(receipt, status)
        raise

File: launcher/lib/asset-preview.mjs
/** Preview is an isolated inspection, never a project job, pin or rights grant. */
import fs from 'node:fs/promises';
import path from 'node:path';
import {randomUUID} from 'node:crypto';
import {assert, fileHash, json, safe, snapshot, writeJson} from './storage.mjs';

export const previewMember = file => /\.(blend|gltf|glb|fbx|bvh)$/i.test(file);
export const previewLimit = 512*1024*1024;

export async function openAssetPreview(work,id,sceneId,revision,request) {
  assert(request&&Object.keys(request).every(k=>['kind','id','version','file'].includes(k)), 'Unknown preview request fields.');
  assert(['catalog','source'].includes(request.kind),'Choose a catalog asset or original package.');
  assert(typeof request.file==='string'&&previewMember(request.file),'Preview supports blend, glTF, GLB, FBX and BVH. Other formats need the specialist workflow.');
  const project=await work.project(id,revision);work.scene(project,sceneId);
  // The project is read only. Preview can coexist with a scene task/candidate.
  let title,root,files,file,motion={};
  if(request.kind==='catalog') {
    const asset=await work.catalogDetail(id,request.id);
    assert(asset.version===request.version,'This asset version changed; reopen Details.',409);
    assert(asset.files.some(f=>f.path===request.file),'Choose an exact recorded asset member.');
    title=asset.title;root=work.config.library;files=asset.files;file=request.file;
    if(asset.kind==='animation') {
      const m=asset.metadata;
      assert(m?.file?.path===file&&m.action&&m.source_object&&m.fps,'Choose the indexed native take member; an unindexed animation needs inspection first.',409);
      motion=Object.fromEntries(['file','action','slot','source_object','fps','frame_start','frame_end'].filter(k=>m[k]!==undefined).map(k=>[k,m[k]]));
    } else assert(['model','pack'].includes(asset.kind),'Material/HDRI preview requires the reviewed look-development workflow.');
  } else {
    const source=await work.sourceDetail(id,request.id);
    assert(source.available&&source.version===request.version,'Package changed or is unavailable; refresh the library.',409);
    assert(source.entrypoints.includes(request.file),'Choose a recorded package member.');
    const version=await json(await safe(work.store.registry,`versions/${source.id}/${source.version}.json`));
    assert(version.sourceId===source.id&&version.version===source.version&&version.relative===source.relative,'Package snapshot identity differs.',409);
    assert(version.files.length<=4096&&version.bytes<=previewLimit,'Preview copy exceeds 4096 files or 512 MiB; use a smaller reviewed package.');
    const base=await safe(work.store.database,source.relative),directory=(await fs.stat(base)).isDirectory();
    assert((await snapshot(base)).version===source.version,'Original package changed; refresh the library.',409);
    root=directory?base:path.dirname(base);files=version.files;title=source.name;
    file=directory?request.file.slice(source.relative.length+1):path.basename(base);
    assert(!directory||request.file.startsWith(source.relative+'/'),'Member leaves its package.');
  }
  assert(files.length<=4096&&files.reduce((n,f)=>n+f.size,0)<=previewLimit,'Preview copy exceeds 4096 files or 512 MiB; use a smaller reviewed package.');
  const previewId='preview_'+randomUUID();
  const base=await safe(work.store.root,'SystemRuntime/UserData/AssetPreviews');
  await fs.mkdir(base,{recursive:true});
  const directory=await safe(base,previewId);await fs.mkdir(directory);
  const requestFile=path.join(directory,'request.json');
  await writeJson(requestFile,{schema:'asset-director.asset-preview/1',id:request.id,title,
    version:request.version,source_kind:request.kind,root,files,file,motion});
  try {
    const receipt=await work.runtime.harness(['workbench-preview','--request',requestFile,'--blender',work.config.blender],205000);
    assert(receipt.state==='READY'&&receipt.source_id===request.id&&receipt.source_version===request.version,'Preview preparation identity mismatch.',409);
    const blend=await safe(directory,receipt.blend.path),bytes=await fileHash(blend);
    assert(bytes.sha256===receipt.blend.sha256&&bytes.size===receipt.blend.size,'Prepared preview changed.',409);
    const launched=await work.runtime.launchAssetPreview(directory,receipt);
    return {previewId,...launched,sourceId:request.id,version:request.version,
      objects:receipt.data.objects.length,takes:receipt.data.takes.length,
      message:'Preview copy opened in a separate Blender window. Nothing was selected or imported. Close that window with X when finished.'};
  } catch(error) {
    // Retain both the native receipt/log and launcher failure; never fabricate READY.
    await writeJson(path.join(directory,'launch-failure.json'),{previewId,state:'FAILED',error:error.message,sourceId:request.id,version:request.version});
    throw new Error(error.message+' Preview attempt retained: '+previewId);
  }
}
